micro_http/src/common/headers.rs

```diff
@@ -13,63 +12,220 @@ pub enum Header {
+                            "identity; q=0" => Err(RequestError::InvalidRequest),
+                            _ => Err(RequestError::InvalidHeader),
+                        },
+                        Header::Expect => match entry[1].trim() {
+                            "100-continue" => {
+                                self.expect = true;
+                                Ok(())
+                            }
+                            _ => Err(RequestError::InvalidRequest),
+                        },
+                    }
+                } else {
+                    Err(RequestError::InvalidHeader)
+                }
+            }
+            _ => Err(RequestError::InvalidHeader),
+        }
     }
 
-    /// Writes the headers to `buf` using the HTTP specification.
-    pub fn write_all<T: Write>(&self, buf: &mut T) -> Result<(), WriteError> {
-        for (key, val) in &self.headers {
-            buf.write_all(key.raw())?;
-            buf.write_all(&[COLON, SP])?;
-            buf.write_all(&val.as_bytes())?;
-            buf.write_all(&[CR, LF])?;
-        }
+    /// Returns the content length of the body.
+    pub fn content_length(&self) -> i32 {
+        self.content_length
+    }
 
-        // The header section ends with a CRLF.
-        buf.write_all(&[CR, LF])?;
+    /// Returns `true` if the transfer encoding is chunked.
+    #[allow(unused)]
+    pub fn chunked(&self) -> bool {
+        self.chunked
+    }
 
-        Ok(())
+    /// Returns `true` if the client is expecting the code 100.
+    #[allow(unused)]
+    pub fn expect(&self) -> bool {
+        self.expect
+    }
+
+    /// Parses a byte slice into a Headers structure for a HTTP request.
+    ///
+    /// The byte slice is expected to have the following format: </br>
+    ///     * Request Header Lines "<header_line> CRLF"- Optional </br>
+    /// There can be any number of request headers, including none, followed by
+    /// an extra sequence of Carriage Return and Line Feed.
+    /// All header fields are parsed. However, only the ones present in the
+    /// [`Headers`](struct.Headers.html) struct are relevant to us and stored
+    /// for future use.
+    ///
+    /// # Errors
+    /// The function returns InvalidRequest when parsing the byte stream fails.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// extern crate micro_http;
+    /// use micro_http::Headers;
+    ///
+    /// let request_headers = Headers::try_from(b"Content-Length: 55\r\n\r\n");
+    /// ```
+    pub fn try_from(bytes: &[u8]) -> Result<Headers, RequestError> {
+        // Headers must be ASCII, so also UTF-8 valid.
+        if let Ok(text) = std::str::from_utf8(bytes) {
+            let mut headers = Headers::default();
+
+            let indexed_lines = text.split("\r\n");
+            for line in indexed_lines {
+                if line.is_empty() {
+                    break;
+                }
+                if let Err(RequestError::InvalidRequest) =
```

Comment:
The following lines (183-188) can be simplified as follows:
```rust
headers.parse_header_line(line.as_bytes())?;
```

Read more about this simplification [here](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html).

Reply:
In the current design, we want to exit the function and return `InvalidRequest` (which invalidates the whole request) only when `parse_header_lines` returns that. Note that `parse_header_lines` can also return `InvalidHeader`, in which case the header line will be discarded but the request would still be valid.

The simplification with `?` would mean exiting the function even on invalid headers, which is not the desired outcome.